File: poc-csharp/RpsPoc/Domain/Move.cs
namespace RpsPoc.Domain;

/// <summary>Jogada de RPS. Camada de dominio: sem thread, sem GUI.</summary>
public enum Move
{
    Rock,
    Paper,
    Scissors,
}

public static class Moves
{
    /// <summary>Jogada aleatoria uniforme (pergunta aberta #4 do planejamento).</summary>
    public static Move Random() => (Move)System.Random.Shared.Next(3);

    /// <returns>1 se <paramref name="self"/> vence, -1 se perde, 0 se empata.</returns>
    public static int Beats(this Move self, Move other)
    {
        if (self == other)
        {
            return 0;
        }

        return self switch
        {
            Move.Rock => other == Move.Scissors ? 1 : -1,
            Move.Paper => other == Move.Rock ? 1 : -1,
            Move.Scissors => other == Move.Paper ? 1 : -1,
            _ => 0,
        };
    }
}

File: poc-java/src/main/java/rps/domain/Move.java
package rps.domain;

import java.util.concurrent.ThreadLocalRandom;

/** Jogada de RPS. Camada de dominio: sem thread, sem GUI. */
public enum Move {
    ROCK,
    PAPER,
    SCISSORS;

    private static final Move[] VALUES = values();

    /** Jogada aleatoria uniforme (pergunta aberta #4 do planejamento). */
    public static Move random() {
        return VALUES[ThreadLocalRandom.current().nextInt(VALUES.length)];
    }

    /**
     * @return 1 se esta jogada vence {@code other}, -1 se perde, 0 se empata.
     */
    public int beats(Move other) {
        if (this == other) {
            return 0;
        }
        return switch (this) {
            case ROCK -> other == SCISSORS ? 1 : -1;
            case PAPER -> other == ROCK ? 1 : -1;
            case SCISSORS -> other == PAPER ? 1 : -1;
        };
    }
}
